### ISSUE TITLE
Flush SQL cache on application exit

### ISSUE THREAD

**Issue by powerman** [nats-io/nats-streaming-server#619]
If I restart docker container with nats-streaming:0.10.2 my clients with durable subscription receive several last (already acked) messages once again, which is not expected behaviour with `stan.MaxInflight(1)`. Looks like it's because SQL cache not flushed on application exit.

**Comment by kozlovic** [comment]
We trap only SIGINT, not SIGTERM. On SIGINT, the store is closed (which causes flush to occur). Will add the trap. Thanks!

**Comment by powerman** [comment]
When it run using `docker run -i -t --rm … nats-streaming:0.10.2 … -SDV` and interrupted using Ctrl-C this doesn't happens, so either in this case docker doesn't delivers SIGINT or it's not handled correctly.

**Comment by kozlovic** [comment]
What do you mean when your Ctrl+C it does not happen? Are you saying that it works as expected? I did verify that we flush the cache on Shutdown. If you Ctrl+C the docker image when running in interactive mode you can see the "Shutdown" message. If you do a `docker stop` however, it is sending the SIGTERM signal that we currently do not trap and there you would not see the "Shutdown" message (and no store flush).

**Comment by powerman** [comment]
I mean after Ctrl-C and then starting it again my test consumer has received once again several last already received and acked (before Ctrl-C) messages. As I'm subscribing with `stan.MaxInflight(1)` this wasn't expected behaviour. So, after I've noticed this I begin looking for other options and found --sql_no_caching, which is expected to solve this issue in other way. But as you have said SIGINT already handled then my consumer shouldn't get these already processed and acked messages (except maybe last one).

**Comment by powerman** [comment]
Here is example session. I've started STAN, consumer and producer. Producer send 10 messages, one per 3 second. Consumer received and acked 10 messages. Then I've pressed Ctrl-C:
```
$ docker run -i -t --rm --name=nss -p 4222:4222 --link=mysql nats-streaming:0.10.2 --cluster_id=dev --store=SQL --sql_driver=mysql --sql_source='root:@tcp(mysql)/nss' -SDV
[1] 2018/07/13 15:29:32.531902 [INF] STREAM: Starting nats-streaming-server[dev] version 0.10.2
[1] 2018/07/13 15:29:32.531934 [INF] STREAM: ServerID: oRxFdtZthCBO0C9ZrOFwhe
[1] 2018/07/13 15:29:32.531938 [INF] STREAM: Go version: go1.10.3
[1] 2018/07/13 15:29:32.536548 [INF] Starting nats-server version 1.1.0
[1] 2018/07/13 15:29:32.536566 [INF] Git commit [not set]
[1] 2018/07/13 15:29:32.537565 [INF] Listening for client connections on 0.0.0.0:4222
[1] 2018/07/13 15:29:32.537570 [INF] Server is ready
[1] 2018/07/13 15:29:32.561712 [TRC] STREAM:  NATS conn opts: { [nats://127.0.0.1:4222] false _NSS-dev-send false false false <nil> true -1 250ms 2s 0s 2m0s 2 0x8e49b0 0x8e4a00 0x8e4960 <nil> 0x8e48f0 -1 8192    <nil> <nil> false}
[1] 2018/07/13 15:29:32.562072 [TRC] STREAM:  NATS conn opts: { [nats://127.0.0.1:4222] false _NSS-dev-general false false false <nil> true -1 250ms 2s 0s 2m0s 2 0x8e49b0 0x8e4a00 0x8e4960 <nil> 0x8e48f0 -1 8192    <nil> <nil> false}
[1] 2018/07/13 15:29:32.562363 [TRC] STREAM:  NATS conn opts: { [nats://127.0.0.1:4222] false _NSS-dev-acks false false false <nil> true -1 250ms 2s 0s 2m0s 2 0x8e49b0 0x8e4a00 0x8e4960 <nil> 0x8e48f0 -1 8192    <nil> <nil> false}
[1] 2018/07/13 15:29:32.562607 [INF] STREAM: Recovering the state...
[1] 2018/07/13 15:29:32.565793 [INF] STREAM: Recovered 2 channel(s)
[1] 2018/07/13 15:29:32.816061 [DBG] STREAM: Did not detect another server instance
[1] 2018/07/13 15:29:32.816095 [DBG] STREAM: Discover subject:           _STAN.discover.dev
[1] 2018/07/13 15:29:32.816099 [DBG] STREAM: Publish subject:            _STAN.pub.MtkdFe5H6IamTywMk7i7ma.>
[1] 2018/07/13 15:29:32.816102 [DBG] STREAM: Subscribe subject:          _STAN.sub.MtkdFe5H6IamTywMk7i7ma
[1] 2018/07/13 15:29:32.816105 [DBG] STREAM: Subscription Close subject: _STAN.subclose.MtkdFe5H6IamTywMk7i7ma
[1] 2018/07/13 15:29:32.816107 [DBG] STREAM: Unsubscribe subject:        _STAN.unsub.MtkdFe5H6IamTywMk7i7ma
[1] 2018/07/13 15:29:32.816110 [DBG] STREAM: Close subject:              _STAN.close.MtkdFe5H6IamTywMk7i7ma
[1] 2018/07/13 15:29:32.816246 [INF] STREAM: Message store is SQL
[1] 2018/07/13 15:29:32.816293 [INF] STREAM: ---------- Store Limits ----------
[1] 2018/07/13 15:29:32.816298 [INF] STREAM: Channels:                  100 *
[1] 2018/07/13 15:29:32.816301 [INF] STREAM: --------- Channels Limits --------
[1] 2018/07/13 15:29:32.816306 [INF] STREAM:   Subscriptions:          1000 *
[1] 2018/07/13 15:29:32.816308 [INF] STREAM:   Messages     :       1000000 *
[1] 2018/07/13 15:29:32.816311 [INF] STREAM:   Bytes        :     976.56 MB *
[1] 2018/07/13 15:29:32.816313 [INF] STREAM:   Age          :     unlimited *
[1] 2018/07/13 15:29:32.816317 [INF] STREAM:   Inactivity   :     unlimited *
[1] 2018/07/13 15:29:32.816320 [INF] STREAM: ----------------------------------
[1] 2018/07/13 15:30:20.218048 [DBG] STREAM: [Client:nss-test-pub] Closed (Inbox=_INBOX.9E49YMtrYnYTgozFFXx2uI)
[1] 2018/07/13 15:30:20.221939 [DBG] STREAM: [Client:nss-test-pub] Replaced old client (Inbox=_INBOX.q6FzkwUK5BL0t1ERcN8WTT)
[1] 2018/07/13 15:30:22.715578 [TRC] STREAM: [Client:nss-test-sub] Delivering msg to subid=26, subject=test, seq=245
[1] 2018/07/13 15:30:22.715940 [TRC] STREAM: [Client:nss-test-sub] Processing ack for subid=26, subject=test, seq=245
[1] 2018/07/13 15:30:22.719893 [TRC] STREAM: [Client:nss-test-pub] Acking Publisher subj=test guid=FpIbXHQOGl1zL4Rdy69LLM
[1] 2018/07/13 15:30:25.727375 [TRC] STREAM: [Client:nss-test-sub] Delivering msg to subid=26, subject=test, seq=246
[1] 2018/07/13 15:30:25.727624 [TRC] STREAM: [Client:nss-test-sub] Processing ack for subid=26, subject=test, seq=246
[1] 2018/07/13 15:30:25.729790 [TRC] STREAM: [Client:nss-test-pub] Acking Publisher subj=test guid=FpIbXHQOGl1zL4Rdy69LLu
[1] 2018/07/13 15:30:28.737097 [TRC] STREAM: [Client:nss-test-sub] Delivering msg to subid=26, subject=test, seq=247
[1] 2018/07/13 15:30:28.737375 [TRC] STREAM: [Client:nss-test-sub] Processing ack for subid=26, subject=test, seq=247
[1] 2018/07/13 15:30:28.739599 [TRC] STREAM: [Client:nss-test-pub] Acking Publisher subj=test guid=FpIbXHQOGl1zL4Rdy69LMS
[1] 2018/07/13 15:30:31.747462 [TRC] STREAM: [Client:nss-test-sub] Delivering msg to subid=26, subject=test, seq=248
[1] 2018/07/13 15:30:31.747728 [TRC] STREAM: [Client:nss-test-sub] Processing ack for subid=26, subject=test, seq=248
[1] 2018/07/13 15:30:31.749962 [TRC] STREAM: [Client:nss-test-pub] Acking Publisher subj=test guid=FpIbXHQOGl1zL4Rdy69LN0
[1] 2018/07/13 15:30:34.759053 [TRC] STREAM: [Client:nss-test-sub] Delivering msg to subid=26, subject=test, seq=249
[1] 2018/07/13 15:30:34.759354 [TRC] STREAM: [Client:nss-test-sub] Processing ack for subid=26, subject=test, seq=249
[1] 2018/07/13 15:30:34.761507 [TRC] STREAM: [Client:nss-test-pub] Acking Publisher subj=test guid=FpIbXHQOGl1zL4Rdy69LNY
[1] 2018/07/13 15:30:37.770250 [TRC] STREAM: [Client:nss-test-sub] Delivering msg to subid=26, subject=test, seq=250
[1] 2018/07/13 15:30:37.770540 [TRC] STREAM: [Client:nss-test-sub] Processing ack for subid=26, subject=test, seq=250
[1] 2018/07/13 15:30:37.777116 [TRC] STREAM: [Client:nss-test-pub] Acking Publisher subj=test guid=FpIbXHQOGl1zL4Rdy69LO6
[1] 2018/07/13 15:30:40.786231 [TRC] STREAM: [Client:nss-test-sub] Delivering msg to subid=26, subject=test, seq=251
[1] 2018/07/13 15:30:40.786487 [TRC] STREAM: [Client:nss-test-sub] Processing ack for subid=26, subject=test, seq=251
[1] 2018/07/13 15:30:40.790248 [TRC] STREAM: [Client:nss-test-pub] Acking Publisher subj=test guid=FpIbXHQOGl1zL4Rdy69LOe
[1] 2018/07/13 15:30:43.798851 [TRC] STREAM: [Client:nss-test-sub] Delivering msg to subid=26, subject=test, seq=252
[1] 2018/07/13 15:30:43.799150 [TRC] STREAM: [Client:nss-test-sub] Processing ack for subid=26, subject=test, seq=252
[1] 2018/07/13 15:30:43.802720 [TRC] STREAM: [Client:nss-test-pub] Acking Publisher subj=test guid=FpIbXHQOGl1zL4Rdy69LPC
[1] 2018/07/13 15:30:46.810158 [TRC] STREAM: [Client:nss-test-sub] Delivering msg to subid=26, subject=test, seq=253
[1] 2018/07/13 15:30:46.810432 [TRC] STREAM: [Client:nss-test-sub] Processing ack for subid=26, subject=test, seq=253
[1] 2018/07/13 15:30:46.812742 [TRC] STREAM: [Client:nss-test-pub] Acking Publisher subj=test guid=FpIbXHQOGl1zL4Rdy69LPk
[1] 2018/07/13 15:30:49.820149 [TRC] STREAM: [Client:nss-test-sub] Delivering msg to subid=26, subject=test, seq=254
[1] 2018/07/13 15:30:49.820424 [TRC] STREAM: [Client:nss-test-sub] Processing ack for subid=26, subject=test, seq=254
[1] 2018/07/13 15:30:49.822678 [TRC] STREAM: [Client:nss-test-pub] Acking Publisher subj=test guid=FpIbXHQOGl1zL4Rdy69LQI
^C[1] 2018/07/13 15:30:51.300641 [INF] STREAM: Shutting down.
[1] 2018/07/13 15:30:51.300930 [DBG] STREAM: connection "_NSS-dev-send" has been closed
[1] 2018/07/13 15:30:51.300946 [DBG] STREAM: connection "_NSS-dev-general" has been closed
[1] 2018/07/13 15:30:51.300958 [DBG] STREAM: connection "_NSS-dev-acks" has been closed
[1] 2018/07/13 15:30:51.300985 [INF] Server Exiting..
```
Now I've also killed producer, while consumer begin auto-reconnection attempts, and I've started STAN once again - and consumer gets 10 last messages redelivered:
```
$ docker run -i -t --rm --name=nss -p 4222:4222 --link=mysql nats-streaming:0.10.2 --cluster_id=dev --store=SQL --sql_driver=mysql --sql_source='root:@tcp(mysql)/nss' -SDV
[1] 2018/07/13 15:31:08.360776 [INF] STREAM: Starting nats-streaming-server[dev] version 0.10.2
[1] 2018/07/13 15:31:08.360804 [INF] STREAM: ServerID: hfUfUAe6QvCHJSzAfx6zIp
[1] 2018/07/13 15:31:08.360809 [INF] STREAM: Go version: go1.10.3
[1] 2018/07/13 15:31:08.364862 [INF] Starting nats-server version 1.1.0
[1] 2018/07/13 15:31:08.364880 [INF] Git commit [not set]
[1] 2018/07/13 15:31:08.365808 [INF] Listening for client connections on 0.0.0.0:4222
[1] 2018/07/13 15:31:08.365813 [INF] Server is ready
[1] 2018/07/13 15:31:08.390012 [TRC] STREAM:  NATS conn opts: { [nats://127.0.0.1:4222] false _NSS-dev-send false false false <nil> true -1 250ms 2s 0s 2m0s 2 0x8e49b0 0x8e4a00 0x8e4960 <nil> 0x8e48f0 -1 8192    <nil> <nil> false}
[1] 2018/07/13 15:31:08.390482 [TRC] STREAM:  NATS conn opts: { [nats://127.0.0.1:4222] false _NSS-dev-general false false false <nil> true -1 250ms 2s 0s 2m0s 2 0x8e49b0 0x8e4a00 0x8e4960 <nil> 0x8e48f0 -1 8192    <nil> <nil> false}
[1] 2018/07/13 15:31:08.390698 [TRC] STREAM:  NATS conn opts: { [nats://127.0.0.1:4222] false _NSS-dev-acks false false false <nil> true -1 250ms 2s 0s 2m0s 2 0x8e49b0 0x8e4a00 0x8e4960 <nil> 0x8e48f0 -1 8192    <nil> <nil> false}
[1] 2018/07/13 15:31:08.390872 [INF] STREAM: Recovering the state...
[1] 2018/07/13 15:31:08.393935 [INF] STREAM: Recovered 2 channel(s)
[1] 2018/07/13 15:31:08.644113 [DBG] STREAM: Did not detect another server instance
[1] 2018/07/13 15:31:08.644170 [DBG] STREAM: Discover subject:           _STAN.discover.dev
[1] 2018/07/13 15:31:08.644177 [DBG] STREAM: Publish subject:            _STAN.pub.MtkdFe5H6IamTywMk7i7ma.>
[1] 2018/07/13 15:31:08.644183 [DBG] STREAM: Subscribe subject:          _STAN.sub.MtkdFe5H6IamTywMk7i7ma
[1] 2018/07/13 15:31:08.644189 [DBG] STREAM: Subscription Close subject: _STAN.subclose.MtkdFe5H6IamTywMk7i7ma
[1] 2018/07/13 15:31:08.644194 [DBG] STREAM: Unsubscribe subject:        _STAN.unsub.MtkdFe5H6IamTywMk7i7ma
[1] 2018/07/13 15:31:08.644200 [DBG] STREAM: Close subject:              _STAN.close.MtkdFe5H6IamTywMk7i7ma
[1] 2018/07/13 15:31:08.644311 [INF] STREAM: Message store is SQL
[1] 2018/07/13 15:31:08.644344 [INF] STREAM: ---------- Store Limits ----------
[1] 2018/07/13 15:31:08.644349 [INF] STREAM: Channels:                  100 *
[1] 2018/07/13 15:31:08.644354 [INF] STREAM: --------- Channels Limits --------
[1] 2018/07/13 15:31:08.644359 [INF] STREAM:   Subscriptions:          1000 *
[1] 2018/07/13 15:31:08.644363 [INF] STREAM:   Messages     :       1000000 *
[1] 2018/07/13 15:31:08.644368 [INF] STREAM:   Bytes        :     976.56 MB *
[1] 2018/07/13 15:31:08.644371 [INF] STREAM:   Age          :     unlimited *
[1] 2018/07/13 15:31:08.644376 [INF] STREAM:   Inactivity   :     unlimited *
[1] 2018/07/13 15:31:08.644379 [INF] STREAM: ----------------------------------
[1] 2018/07/13 15:31:08.644795 [TRC] STREAM: [Client:nss-test-sub] Delivering msg to subid=26, subject=test, seq=245
[1] 2018/07/13 15:31:11.058152 [DBG] STREAM: [Client:nss-test-sub] Suspended durable subscription, subject=test, inbox=_INBOX.iUcFfC7dxX5KZ02QPYU64w, durable=reader, subid=26
[1] 2018/07/13 15:31:11.058164 [DBG] STREAM: [Client:nss-test-sub] Closed (Inbox=_INBOX.iUcFfC7dxX5KZ02QPYU5xv)
[1] 2018/07/13 15:31:11.062230 [DBG] STREAM: [Client:nss-test-sub] Replaced old client (Inbox=_INBOX.iUcFfC7dxX5KZ02QPYU6HZ)
[1] 2018/07/13 15:31:11.069718 [DBG] STREAM: [Client:nss-test-sub] Resumed durable subscription, subject=test, inbox=_INBOX.iUcFfC7dxX5KZ02QPYU6Oa, durable=reader, subid=26
[1] 2018/07/13 15:31:11.069753 [DBG] STREAM: [Client:nss-test-sub] Redelivering to subid=26, durable=reader
[1] 2018/07/13 15:31:11.069922 [TRC] STREAM: [Client:nss-test-sub] Redelivering to subid=26, seq=245
[1] 2018/07/13 15:31:11.069928 [TRC] STREAM: [Client:nss-test-sub] Redelivering msg to subid=26, subject=test, seq=245
[1] 2018/07/13 15:31:11.070172 [TRC] STREAM: [Client:nss-test-sub] Processing ack for subid=26, subject=test, seq=245
[1] 2018/07/13 15:31:11.077479 [TRC] STREAM: [Client:nss-test-sub] Delivering msg to subid=26, subject=test, seq=246
[1] 2018/07/13 15:31:11.077728 [TRC] STREAM: [Client:nss-test-sub] Processing ack for subid=26, subject=test, seq=246
[1] 2018/07/13 15:31:11.077832 [TRC] STREAM: [Client:nss-test-sub] Delivering msg to subid=26, subject=test, seq=247
[1] 2018/07/13 15:31:11.077984 [TRC] STREAM: [Client:nss-test-sub] Processing ack for subid=26, subject=test, seq=247
[1] 2018/07/13 15:31:11.078091 [TRC] STREAM: [Client:nss-test-sub] Delivering msg to subid=26, subject=test, seq=248
[1] 2018/07/13 15:31:11.078240 [TRC] STREAM: [Client:nss-test-sub] Processing ack for subid=26, subject=test, seq=248
[1] 2018/07/13 15:31:11.078344 [TRC] STREAM: [Client:nss-test-sub] Delivering msg to subid=26, subject=test, seq=249
[1] 2018/07/13 15:31:11.078476 [TRC] STREAM: [Client:nss-test-sub] Processing ack for subid=26, subject=test, seq=249
[1] 2018/07/13 15:31:11.078570 [TRC] STREAM: [Client:nss-test-sub] Delivering msg to subid=26, subject=test, seq=250
[1] 2018/07/13 15:31:11.078704 [TRC] STREAM: [Client:nss-test-sub] Processing ack for subid=26, subject=test, seq=250
[1] 2018/07/13 15:31:11.078796 [TRC] STREAM: [Client:nss-test-sub] Delivering msg to subid=26, subject=test, seq=251
[1] 2018/07/13 15:31:11.078929 [TRC] STREAM: [Client:nss-test-sub] Processing ack for subid=26, subject=test, seq=251
[1] 2018/07/13 15:31:11.079030 [TRC] STREAM: [Client:nss-test-sub] Delivering msg to subid=26, subject=test, seq=252
[1] 2018/07/13 15:31:11.079155 [TRC] STREAM: [Client:nss-test-sub] Processing ack for subid=26, subject=test, seq=252
[1] 2018/07/13 15:31:11.079256 [TRC] STREAM: [Client:nss-test-sub] Delivering msg to subid=26, subject=test, seq=253
[1] 2018/07/13 15:31:11.079396 [TRC] STREAM: [Client:nss-test-sub] Processing ack for subid=26, subject=test, seq=253
[1] 2018/07/13 15:31:11.079498 [TRC] STREAM: [Client:nss-test-sub] Delivering msg to subid=26, subject=test, seq=254
[1] 2018/07/13 15:31:11.079634 [TRC] STREAM: [Client:nss-test-sub] Processing ack for subid=26, subject=test, seq=254
```

**Comment by powerman** [comment]
Here is consumer's code, just in case:
```go
package main

import (
	"time"

	"github.com/nats-io/go-nats-streaming"
	"github.com/powerman/structlog"
)

var log = structlog.New()

const clusterID = "dev"
const clientID = "nss-test"

func main() {
	structlog.DefaultLogger.SetPrefixKeys(structlog.KeyTime, structlog.KeyLevel).SetSuffixKeys(structlog.KeyStack)
	go stanSubscribe("nats://localhost:4222", clusterID, clientID, func(sc stan.Conn) error {
		_, err := sc.Subscribe("test", func(m *stan.Msg) {
			log.Info("received", "msg", string(m.Data))
		},
			stan.DeliverAllAvailable(),
			stan.DurableName("reader"),
			stan.MaxInflight(1),
		)
		return err
	})

	select {}
}

func stanSubscribe(natsURL, stanClusterID, clientID string, subscribe func(stan.Conn) error) {
	const minDelay = time.Second / 10
	const maxDelay = 5 * time.Second
	delay := minDelay
	connlost := make(chan error)
	for {
		sc, err := stan.Connect(stanClusterID, clientID+"-sub",
			stan.NatsURL(natsURL),
			stan.SetConnectionLostHandler(func(_ stan.Conn, err error) {
				connlost <- err
			}),
		)
		if err != nil {
			log.Err("failed to connect", "err", err)
			time.Sleep(delay)
			delay *= 2
			if delay > maxDelay {
				delay = maxDelay
			}
			continue
		}
		delay = minDelay
		err = subscribe(sc)
		if err != nil {
			log.Err("failed to subscribe", "err", err)
			log.WarnIfFail(sc.Close)
			continue
		}
		err = <-connlost
		log.Err("connection lost", "err", err)
		log.WarnIfFail(sc.Close)
	}
}
```

**Comment by kozlovic** [comment]
Thanks! Having a look now.

**Comment by kozlovic** [comment]
Ok, I think I found what the problem is. Should have a fix for it today. Thanks for the report!

**Comment by kozlovic** [comment]
This should be fixed with PR #627. Again, thank you for the report!
